### PR TITLE
Indirect Elwin - Obtain Units from Sample Logs

### LINKED ISSUE #16019: Indirect Elwin - Obtain unit from sample logs
The Elwin interface/algorithm should automatically obtain the correct unit for the input workspace by checking the sample logs for the workspace

### PR DESCRIPTION
**Pull request by louisemccann**
The elastic window multiple now reads the sample environment units from the sample logs instead of just assuming that the sample environment is in temperature.

**To test:**

`\\olympic\babylon5\Public\Louise McCann\DataAnalysis\Elwin`

Interfaces>Data Analysis>Elwin

Input Files: `osiris120225_graphite002_red`, `osiris120226_graphite002_red`, `osiris120227_graphite002_red`

SE log name: `Field`
Press `Run`
Once algorithm complete, plot workspace with the suffix `_elf`
The x-axis should be labeled `Field` and have units of Tesla (T)

Repeat with SE log name: `sample`
x-axis units should be in temperature (K)

Fixes #16019.

[Release Notes](https://github.com/mantidproject/mantid/blob/2b6eb15cb7dd9bbc22efcc02aacc444793036761/docs/source/release/v3.9.0/indirect_inelastic.rst)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py [hunk 1]
```diff
@@ -99,41 +98,50 @@ def PyExec(self):
...
-
-            Logarithm(InputWorkspace=q2_ws,
-                      OutputWorkspace=q2_ws)
+                elwin_alg.setAlwaysStoreInADS(True)
+                elwin_alg.setProperty("InputWorkspace", input_ws)
+                elwin_alg.setProperty("IntegrationRangeStart", self._integration_range_start)
+                elwin_alg.setProperty("IntegrationRangeEnd", self._integration_range_end)
+                elwin_alg.setProperty("OutputInQ", q_ws)
+                elwin_alg.setProperty("OutputInQSquared", q2_ws)
+                elwin_alg.execute()
```
Comment (PranavBahuguna) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py]: Why not just do something like this?

```elwin_alg = self.createChildAlgorithm("ElasticWindow", enableLogging=False)
elwin_alg.setAlwaysStoreInADS(True)
elwin_alg.setProperty("InputWorkspace", input_ws)
elwin_alg.setProperty("IntegrationRangeStart", self._integration_range_start)
elwin_alg.setProperty("IntegrationRangeEnd", self._integration_range_end)
elwin_alg.setProperty("OutputInQ", q_ws)
elwin_alg.setProperty("OutputInQSquared", q2_ws)

if self._background_range_start != Property.EMPTY_DBL and self._background_range_end != Property.EMPTY_DBL:
    elwin_alg.setProperty("BackgroundRangeStart", self._background_range_start)
    elwin_alg.setProperty("BackgroundRangeEnd", self._background_range_end)

elwin_alg.execute()
```

**Comment by PranavBahuguna** [comment]
Following the testing instructions produces the expected results. The code itself (that you wrote) also seems fine, just see my line comments and try to make improvements where one could obviously be made.

**Comment by PranavBahuguna** [comment]
Alright, all's good :shipit:

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py [hunk 2]
```diff
@@ -99,41 +98,44 @@ def PyExec(self):
...
         for input_ws in input_workspace_names:
+
             logger.information('Running ElasticWindow for workspace: %s' % input_ws)
+            progress.report('ElasticWindow for workspace: %s' % input_ws)
 
             q_ws = '__' + input_ws + '_q'
             q2_ws = '__' + input_ws + '_q2'
 
+            elwin_alg = self.createChildAlgorithm("ElasticWindow", enableLogging=False)
+            elwin_alg.setAlwaysStoreInADS(True)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py]: It's good that we've switched to the `createChildAlgorithm` style. It's better practice if algorithms don't interact with the ADS themselves. Do we need this line?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py [hunk 3]
```diff
@@ -142,54 +144,68 @@ def PyExec(self):
...
+            append_alg.setProperty("InputWorkspace1", q_workspaces[0])
+            append_alg.setProperty("InputWorkspace2", q_workspaces[1])
+            append_alg.setProperty("OutputWorkspace", self._q_workspace)
+            append_alg.execute()
+            mtd.addOrReplace(self._q_workspace, append_alg.getProperty("OutputWorkspace").value)
+            append_alg.setProperty("InputWorkspace1", q2_workspaces[0])
+            append_alg.setProperty("InputWorkspace2", q2_workspaces[1])
+            append_alg.setProperty("OutputWorkspace", self._q2_workspace)
+            append_alg.execute()
+            mtd.addOrReplace(self._q2_workspace, append_alg.getProperty("OutputWorkspace").value)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py]: Similar comment. We have the output properties and setting those will control if the workspace goes into the ADS so we don't and shouldn't add it manually.

I can see there are a few places where this type of line is done.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py [hunk 4]
```diff
@@ -142,54 +144,68 @@ def PyExec(self):
...
+                append_alg.execute()
+                mtd.addOrReplace(self._q_workspace, append_alg.getProperty("OutputWorkspace").value)
+                append_alg.setProperty("InputWorkspace1", self._q2_workspace)
+                append_alg.setProperty("InputWorkspace2", q2_workspaces[idx])
+                append_alg.setProperty("OutputWorkspace", self._q2_workspace)
+                append_alg.execute()
+                mtd.addOrReplace(self._q2_workspace, append_alg.getProperty("OutputWorkspace").value)
 
             # Delete the output workspaces from the ElasticWindow algorithms
+            delete_alg = self.createChildAlgorithm("DeleteWorkspace", enableLogging=False)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py]: If we remove the `mtd.add` lines then we shouldn't need to delete the workspaces at all as they will not be visible externally.

**Comment by louisemccann** [comment]
For some reason SortXAxis wouldn't work when the input workspace wasn't in the ADS

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py [hunk 5]
```diff
@@ -227,26 +220,25 @@ def PyExec(self):
...
 
         # Add the new vertical axis to each workspace
-        mtd[self._q_workspace].replaceAxis(1, q_ws_axis)
-        mtd[self._q2_workspace].replaceAxis(1, q2_ws_axis)
+        self._q_workspace.replaceAxis(1, q_ws_axis)
+        self._q2_workspace.replaceAxis(1, q2_ws_axis)
 
         progress.report('Creating ELF workspaces')
         transpose_alg = self.createChildAlgorithm("Transpose", enableLogging=False)
+        transpose_alg.setAlwaysStoreInADS(True)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ElasticWindowMultiple.py]: Is this line necessary? It seems like the only line left hitting the ADS.

**Comment by louisemccann** [comment]
Removing that line and using `sort_alg.setProperty("InputWorkspace", transpose_alg.getProperty("OutputWorkspace").value)`
I get the following error:
`RuntimeError: Invalid value for property InputWorkspace (Workspace) "": Enter a name for the Input/InOut workspace
  at line 32 in 'C:/Users/PKB22426/Documents/Build1/mantid/Framework/PythonInterface/plugins/algorithms\SortXAxis.py'
  caused by line 924 in 'C:\Users\PKB22426\Documents\Build1\Build\bin\Release\mantid\simpleapi.py'
  caused by line 870 in 'C:\Users\PKB22426\Documents\Build1\Build\bin\Release\mantid\simpleapi.py'
  caused by line 860 in 'C:\Users\PKB22426\Documents\Build1\Build\bin\Release\mantid\simpleapi.py'`

**Comment by martyngigg** [comment]
@louisemccann I see the problem. [`SortX`](/mantidproject/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/SortXAxis.py) is actually the issue (line 32). It doesn't use the `createChild` algorithm approach so requires a name and the ADS.

I'm conscious that this is then going beyond this algorithm and too much change for now. The changes you have already made are a good improvement. Could you create an issue to clean that line and `SortXAxis` for maintenance?

**Comment by martyngigg** [comment]
@louisemccann Unfortunatley it looks like there is a failing system test now.

**Comment by martyngigg** [comment]
I've tried to get this through the system test job a couple of times but other problems have prevented it. I have run all of the ISISIndirect system tests on this and they are all okay so I'm happy.

**Pull request merged**